# Dataset categories

Exploring the available cyberattack and malware categories available on each dataset.

In [1]:
import pandas as pd
import os

def show_categories(column_name: str, csv_file: str, clipboard: bool = False):
    df = pd.read_csv(csv_file, usecols=[column_name])
    samples_per_category = df[column_name].value_counts()
    if clipboard:
        samples_per_category.to_clipboard()
    print(f"Total rows: {df.shape[0]}")
    print(samples_per_category)

DATASET_DIRECTORY = "/mnt/btrfs/datasets/cybersecurity/multiclass"

## After malware category standardization
### Unified

In [ ]:
dataset_file = os.path.join(DATASET_DIRECTORY, "unified", "cybersecurity-unified-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 62537257
Label
Normal         21662749
DoS            13063169
Scanner        12630622
Botnet          9362114
Brute force     3768782
Generic         1746551
Exploit          250465
MITM              36038
Infection         16767
Name: count, dtype: int64


As a result, if we want to perform balanced multiclass malware classification, we will need to have $16767$ samples for each of the $9$ availables classes, resulting in a total of $16767 \times 9 = 150903$ samples from the unified dataset.

### IoT datasets
#### IoT-23

In [ ]:
dataset_file = os.path.join(DATASET_DIRECTORY, "iot", "IoT-23-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 37299318
Label
Scanner      12293306
DoS          10643480
Botnet        8344491
Normal        4280339
Generic       1716778
Exploit         12889
Infection        8035
Name: count, dtype: int64


#### IoT-DS2

In [7]:
dataset_file = os.path.join(DATASET_DIRECTORY, "iot", "IoT-DS2-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 1246536
Label
DoS            437682
Normal         224709
Scanner        198803
Botnet         196259
Brute force    132843
Generic         25141
Exploit         12889
MITM            10176
Infection        8034
Name: count, dtype: int64


#### IotID20

In [8]:
dataset_file = os.path.join(DATASET_DIRECTORY, "iot", "iotid20-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 411673
Label
Botnet     231107
DoS         59390
Scanner     56744
Normal      38570
MITM        25862
Name: count, dtype: int64


#### MQTT-IoT-IDS2020

In [9]:
dataset_file = os.path.join(DATASET_DIRECTORY, "iot", "mqtt-iot-ids2020-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 3486039
Label
Brute force    3254231
Normal          167159
Scanner          64649
Name: count, dtype: int64


### IDS
#### CIC-UNSW-NB15

In [10]:
dataset_file = os.path.join(DATASET_DIRECTORY, "ids", "CIC-UNSW-NB15-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 3480335
Label
Normal       3390752
Exploit        62666
Scanner        17120
Generic         4632
DoS             4467
Infection        698
Name: count, dtype: int64


#### IDS-2018

In [11]:
dataset_file = os.path.join(DATASET_DIRECTORY, "ids", "ids-2018-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 16012791
Label
Normal         13264722
DoS             1918150
Brute force      381708
Botnet           286190
Exploit          162021
Name: count, dtype: int64


#### USTC-TFC-2016

In [12]:
dataset_file = os.path.join(DATASET_DIRECTORY, "ids", "ustc-tfc-2016-multiclass.csv")
show_categories("Label", dataset_file)

Total rows: 616362
Label
Normal    312294
Botnet    304068
Name: count, dtype: int64
